writer.go

```diff
@@ -289,523 +541,531 @@ func NewWriter(config WriterConfig) *Writer {
 // best way to achieve good batching behavior is to share one Writer amongst
 // multiple go routines.
 //
-// When the method returns an error, there's no way to know yet which messages
-// have succeeded of failed.
+// When the method returns an error, it may be of type kafka.WriteError to allow
+// the caller to determine the status of each message.
 //
 // The context passed as first argument may also be used to asynchronously
 // cancel the operation. Note that in this case there are no guarantees made on
 // whether messages were written to kafka. The program should assume that the
 // whole batch failed and re-write the messages later (which could then cause
 // duplicates).
 func (w *Writer) WriteMessages(ctx context.Context, msgs ...Message) error {
-	if len(msgs) == 0 {
-		return nil
+	if w.Addr == nil {
+		return errors.New("kafka.(*Writer).WriteMessages: cannot create a kafka writer with a nil address")
 	}
 
-	var err error
-	var res chan error
-	if !w.config.Async {
-		res = make(chan error, len(msgs))
+	if w.Topic == "" {
+		return errors.New("kafka.(*Writer).WriteMessages: cannot create a kafka writer with an empty topic")
 	}
-	t0 := time.Now()
-	defer w.stats.writeTime.observeDuration(time.Since(t0))
 
-	w.mutex.RLock()
-	closed := w.closed
-	w.mutex.RUnlock()
+	w.group.Add(1)
+	defer w.group.Done()
 
-	if closed {
+	if w.isClosed() {
 		return io.ErrClosedPipe
 	}
 
-	for i, msg := range msgs {
+	if len(msgs) == 0 {
+		return nil
+	}
 
-		if int(msg.size()) > w.config.BatchBytes {
-			err := MessageTooLargeError{
-				Message:   msg,
-				Remaining: msgs[i+1:],
-			}
-			return err
+	balancer := w.balancer()
+	batchBytes := w.batchBytes()
+
+	for i := range msgs {
+		n := int64(msgs[i].size())
+		if n > batchBytes {
+			// This error is left for backward compatibility with historical
+			// behavior, but it can yield O(N^2) behaviors. The expectations
+			// are that the program will check if WriteMessages returned a
+			// MessageTooLargeError, discard the message that was exceeding
+			// the maximum size, and try again.
+			return messageTooLarge(msgs, i)
 		}
+	}
 
-		wm := writerMessage{msg: msg, res: res}
+	numPartitions, err := w.partitions(ctx)
+	if err != nil {
```

Comment:
If I'm reading this right, we would make an expensive call to the broker to fetch the number of partitions for every WriteMessage call.

Reply:
I see that caching is done in the new Transport, maybe a comment would help?